### PR TITLE
PARQUET-2120: Dictionary command should handle missing dictionary pages

### PR DESCRIPTION
**Pull request by rshkv**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Parquet Jira](https://issues.apache.org/jira/browse/PARQUET/) issues and references them in the PR title. For example, "PARQUET-1234: My Parquet PR"
  - https://issues.apache.org/jira/browse/PARQUET-2120
  - ~In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).~

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain Javadoc that explain what it does

### CODE REVIEW

File: parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java [hunk 1]
```diff
@@ -75,40 +75,12 @@ public int run() throws IOException {
...
-              break;
-            case DOUBLE:
-              console.info("{}: {}", String.format("%6d", i),
-                  dict.decodeToDouble(i));
-              break;
-            default:
-              throw new IllegalArgumentException(
-                  "Unknown dictionary type: " + type.getPrimitiveTypeName());
-          }
+        if (page != null) {
```
Comment (rshkv) [inline parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java:78]: This check is the crux of the change.

File: parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java [hunk 2]
```diff
@@ -75,40 +75,12 @@ public int run() throws IOException {
...
+          console.info("\nRow group {} dictionary for \"{}\":", rowGroup, column);
+          Dictionary dict = page.getEncoding().initDictionary(descriptor, page);
+          printDictionary(dict, type);
+        } else {
+          console.info("\nRow group {} has no dictionary for \"{}\"", rowGroup, column);
```
Comment (rshkv) [inline parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java:83]: For a file mixing pages with and without dictionary encoding the output would look e.g. like this:
```
Row group 0 has no dictionary for "col"

Row group 1 dictionary for "col":
     0: "b"
     1: "c"
```

File: parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java [hunk 3]
```diff
@@ -122,6 +94,41 @@ public int run() throws IOException {
     return 0;
   }
 
+  private void printDictionary(Dictionary dict, PrimitiveType type) {
```
Comment (rshkv) [inline parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java:97]: This is just a copy-paste of the `for` block above.

File: parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java [hunk 4]
```diff
@@ -75,40 +75,12 @@ public int run() throws IOException {
       while ((dictionaryReader = reader.getNextDictionaryReader()) != null) {
         DictionaryPage page = dictionaryReader.readDictionaryPage(descriptor);
 
-        Dictionary dict = page.getEncoding().initDictionary(descriptor, page);
-
-        console.info("\nRow group {} dictionary for \"{}\":", rowGroup, column, page.getCompressedSize());
```
Comment (rshkv) [inline parquet-cli/src/main/java/org/apache/parquet/cli/commands/ShowDictionaryCommand.java:80]: I removed the `page.getCompressedSize()` argument here as the log didn't have enough placeholders to display it in the first place.

**Comment by shangxinli** [comment]
Thanks for working on it! Can you squash the commits?

**Comment by rshkv** [comment]
Done!

**Comment by rshkv** [comment]
@shangxinli anything else I can do here?

**Comment by shangxinli** [comment]
I will have a look soon sometime this week.

**Review by shangxinli — APPROVED** [review]
LGTM, Thanks for working on it!

**Comment by rshkv** [comment]
@shangxinli, thanks for reviewing!

**Pull request merged**